cmd/src/debug_kube.go

```diff
@@ -0,0 +1,389 @@
+			wg.Add(1)
+			go func(pod, container string) {
+				if err := semaphore.Acquire(ctx, 1); err != nil {
+					return
+				}
+				defer semaphore.Release(1)
+				defer wg.Done()
+				ch <- getPodLog(ctx, pod, container, namespace, baseDir)
+			}(pod.Metadata.Name, container.Name)
+		}
+	}
+
+	// start goroutine to run kubectl logs --previous for each pod's container's
+	// won't write to zip on err, only passes bytes to channel if err not nil
+	// TODO: It may be nice to store a list of pods for which --previous isn't collected, to be outputted with verbose flag
+	for _, pod := range pods.Items {
+		for _, container := range pod.Spec.Containers {
+			wg.Add(1)
+			go func(pod, container string) {
+				if err := semaphore.Acquire(ctx, 1); err != nil {
+					return
+				}
+				defer semaphore.Release(1)
+				defer wg.Done()
+				f := getPastPodLog(ctx, pod, container, namespace, baseDir)
+				if f.err == nil {
+					ch <- f
+				} else if verbose {
+					fmt.Printf("Could not gather --previous pod logs for: %s \nExited with err: %s\n", pod, f.err)
+				}
+			}(pod.Metadata.Name, container.Name)
+		}
+	}
+
+	// start goroutine for each pod to run kubectl describe pod
+	for _, pod := range pods.Items {
+		wg.Add(1)
+		go func(pod string) {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return
+			}
+			defer semaphore.Release(1)
+			defer wg.Done()
+			ch <- getDescribe(ctx, pod, namespace, baseDir)
+		}(pod.Metadata.Name)
+	}
+
+	// start goroutine for each pod to run kubectl get pod <pod> -o yaml
+	for _, pod := range pods.Items {
+		wg.Add(1)
+		go func(pod string) {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return
+			}
+			defer semaphore.Release(1)
+			defer wg.Done()
+			ch <- getManifest(ctx, pod, namespace, baseDir)
+		}(pod.Metadata.Name)
+	}
+
+	// start goroutine to get external service config
+	if configs {
+		wg.Add(1)
+		go func() {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return
+			}
+			defer semaphore.Release(1)
+			defer wg.Done()
+			ch <- getExternalServicesConfig(ctx, baseDir)
+		}()
+
+		wg.Add(1)
+		go func() {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return
+			}
+			defer semaphore.Release(1)
+			defer wg.Done()
+			ch <- getSiteConfig(ctx, baseDir)
+		}()
+	}
+
+	// close channel when wait group goroutines have completed
+	go func() {
+		wg.Wait()
+		close(ch)
+	}()
+
+	// write to archive all the outputs from kubectl call functions passed to buffer channel
+	for f := range ch {
+		if f.err != nil {
+			log.Printf("getting data for %s failed: %v\noutput: %s", f.name, f.err, f.data)
+			continue
+		}
+
+		if verbose {
+			log.Printf("archiving file %q with %d bytes", f.name, len(f.data))
+		}
+
+		zf, err := zw.Create(f.name)
+		if err != nil {
+			return fmt.Errorf("failed to create %s: %w", f.name, err)
+		}
+
+		_, err = zf.Write(f.data)
+		if err != nil {
+			return fmt.Errorf("failed to write to %s: %w", f.name, err)
+		}
+	}
```

Comment:
This is repeated a few times (without repetition and without needing to cleanup allocated resources) and is a candidate for a named function in debug_common.

Something like `func writeChannelContentsToZip(zw *zip.Writer, ch <-chan *archiveFile) error`

Reply:
Did this ⚡